### ISSUE TITLE
Unclear support status for CMYK in v0.7

### ISSUE THREAD

**Issue by amyspark** [libjxl/libjxl#1719]
**Describe the bug**

Throughout the libjxl code there are references to the CMYK color space being supported. However, 

https://github.com/libjxl/libjxl/blob/4c31ef03e3fd5239d6b794771d4ae7daa7815b28/lib/jxl/encode.cc#L754

this check actively prevents us from exporting JXL CMYK images.

**To Reproduce**
Call JxlEncoderSetBasicInfo with basicInfo.num_color_channels = 4 and num_extra_channels = 1.

**Expected behavior**
Export working out of the box.

**Screenshots**
N/A

**Environment**
 - OS: [e.g. Windows] Windows 10 21H2
 - Compiler version: [e.g. clang 11.0.1] MSVC 19.34.31721
 - CPU type: [e.g. x86_64] x86_64
 - cjxl/djxl version string: [e.g. cjxl [v0.3.7 | SIMD supported: SSE4,Scalar]]  JPEG XL encoder v0.7.0 0.7.0 [AVX2,SSE4,SSSE3,Unknown]

**Additional context**
N/A

**Comment by novomesk** [comment]
I guess when encoding CMYK, you have to use `basicInfo.num_color_channels = 3` (CMY) and `num_extra_channels = 1` for (K).

Extra channel should be probably `JXL_CHANNEL_BLACK`

I haven't tried it yet.

**Comment by 0xC0000054** [comment]
There is a CMYK + Alpha [test image](https://github.com/libjxl/conformance/blob/master/testcases/cmyk_layers/input.jxl) in the libjxl conformance repository.
libjxl shows that image has having 3 color channels (CMY) and 2 extra channels (K and Alpha).

**Comment by sati32** [comment]
That is 5 channels.
This will not work because of
`  if (format->num_channels > 4) {
    return JXL_API_ERROR("More than 4 channels not supported");
`
in decode.cc

**Comment by novomesk** [comment]
I think that `JxlDecoderSetExtraChannelBuffer` and `JxlEncoderSetExtraChannelBuffer` must be used to read/write the K channel.

**Comment by jonsneyers** [comment]
That's right. The single interleaved buffer passing is only defined for grayscale/RGB with or without alpha (so 1 to 4  channels), but if you have any other extra channels than alpha, they need to be passed separately.

**Comment by sati32** [comment]
OK, this way I managed to load the test image in PhotoLine. But it's a confusing code structure for a new file format.

**Comment by mo271** [comment]
Is this mainly a documentation problem, or do we actually want to change behavior here?

**Comment by amyspark** [comment]
The behavior itself is problematic, because SetExtraChannelBuffer only accepts *extra* channels. CMYK is *4* *color* channels together with the alpha channel, there's nothing that could be considered "extra" from the point of view of the color space. As is, I don't yet find a way in which we could support CMYK in JXL because of the complexity of partially interleaving caller-owned plane buffers (plus the resource costs).

I could suggest three alternatives:

- document the behaviour of the function, suggest explicitly how larger spaces could be dealt with
- enable planar buffer upload for *all* channels, not just the so-called "extra" channels
- drop the RGB/grayscale limitation for interleaved buffers (easiest IMO for end users)

**Comment by mo271** [comment]
In some sense this is the encoder analogue to this decoder issue: https://github.com/libjxl/libjxl/issues/1542

**Comment by jonsneyers** [comment]
In the jxl spec, CMYK is treated as CMY as color channels + K as extra channels.
In the libjxl implementation, all channels are planar anyway.

I agree that it would be convenient to generalize the libjxl API to allow passing any channel in a planar way, as well as adding an interleaved option for CMYK and other 4-channel data besides RGBA, e.g. RGBDepth, RGBSelectionmask, RGBThermal etc could also be useful. Possibly also interleaving of more than 4 channels could be added.
